### ISSUE TITLE
Extend null_or_empty to string types

### ISSUE THREAD

**Issue by robd003** [crate/crate#14076]
### Problem Statement

It would be great if I could use null_or_empty for string types too.

Current code:
```
SELECT
  CASE
    WHEN sessions.referrer IS NULL THEN 'Direct'
    WHEN sessions.referrer = '' THEN 'Direct'
    ELSE sessions.referrer
  END as referrer_result,
```

Preferred code:
```
SELECT
  CASE
    WHEN null_or_empty(sessions.referrer) THEN 'Direct'
    ELSE sessions.referrer
  END as referrer_result,
```

However null_or_empty seems to be only looking for objects and fails with:
```
java.lang.RuntimeException: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'https': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (StringReader); line: 1, column: 6]
	at io.crate.types.ObjectType.mapFromJSONString(ObjectType.java:210)
	at io.crate.types.ObjectType.convert(ObjectType.java:176)
	at io.crate.types.ObjectType.implicitCast(ObjectType.java:159)
	at io.crate.types.ObjectType.implicitCast(ObjectType.java:67)
	at io.crate.expression.scalar.cast.ImplicitCastFunction.castToTargetType(ImplicitCastFunction.java:108)
	at io.crate.expression.scalar.cast.ImplicitCastFunction.evaluate(ImplicitCastFunction.java:102)
	at io.crate.expression.FunctionExpression.value(FunctionExpression.java:50)
	at io.crate.expression.scalar.NullOrEmptyFunction.evaluate(NullOrEmptyFunction.java:93)
	at io.crate.expression.scalar.NullOrEmptyFunction.evaluate(NullOrEmptyFunction.java:51)
	at io.crate.expression.FunctionExpression.value(FunctionExpression.java:50)
	at io.crate.expression.scalar.conditional.CaseFunction.evaluate(CaseFunction.java:103)
	at io.crate.expression.FunctionExpression.value(FunctionExpression.java:50)
	at io.crate.execution.engine.aggregation.GroupingCollector.lambda$singleKey$1(GroupingCollector.java:95)
	at io.crate.execution.engine.aggregation.GroupingCollector.iter(GroupingCollector.java:230)
	at io.crate.data.BatchIterators.lambda$collect$0(BatchIterators.java:59)
	at io.crate.data.BatchIterator.move(BatchIterator.java:155)
	at io.crate.data.BatchIterator.lambda$move$0(BatchIterator.java:173)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.crate.data.BatchIterators.lambda$collect$1(BatchIterators.java:62)
	at io.crate.data.BatchIterator.move(BatchIterator.java:158)
	at io.crate.data.BatchIterator.lambda$move$0(BatchIterator.java:173)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.crate.execution.IncrementalPageBucketReceiver.consumeRows(IncrementalPageBucketReceiver.java:129)
	at io.crate.execution.IncrementalPageBucketReceiver.lambda$setBucket$5(IncrementalPageBucketReceiver.java:112)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1810)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'https': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (StringReader); line: 1, column: 6]
	at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2418)
	at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:759)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._reportInvalidToken(ReaderBasedJsonParser.java:3038)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._handleOddValue(ReaderBasedJsonParser.java:2079)
	at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.nextToken(ReaderBasedJsonParser.java:805)
	at org.elasticsearch.common.xcontent.json.JsonXContentParser.nextToken(JsonXContentParser.java:53)
	at org.elasticsearch.common.xcontent.support.AbstractXContentParser.readMap(AbstractXContentParser.java:330)
	at org.elasticsearch.common.xcontent.support.AbstractXContentParser.readMap(AbstractXContentParser.java:303)
	at org.elasticsearch.common.xcontent.support.AbstractXContentParser.map(AbstractXContentParser.java:243)
	at io.crate.types.ObjectType.mapFromJSONString(ObjectType.java:208)
```

### Possible Solutions

Just was looking for a faster way to see if a string was NULL or an empty string.

### Considered Alternatives

Current solution works, but I figured reducing the number of CASE WHEN statements would improve performance.

**Comment by matriv** [comment]
Imho, it doesn't add any substantial value, you can simply do this:
```
SELECT
  CASE
    WHEN sessions.referrer IS NULL OR sessions.referrer = '' THEN 'Direct'
    ELSE sessions.referrer
  END as referrer_result,
```
There is no performance gain here, the `empty_or_null` would need to do the same checks.

Again, imho, `empty_or_null` is a special function to be called on objects, and is not a common function found in traditional RDBMS, so I'd vote against adding support for strings.

**Comment by mfussenegger** [comment]
Since https://github.com/crate/crate/pull/13401 additional case expressions shouldn't have that much of an impact anymore. It was mostly an issue with the tree-representation.

My initial implementation of `null_or_empty` actually did contain string support, but I dropped it once I realized that I'd have to add special query handling, otherwise it would result in _worse_ performance if used in the WHERE clause.

The reason `null_or_empty` helps with performance for objects is because we don't have indices for objects themselves, only for the children, so we have to do extra for `IS NULL` to figure out if it's absent or only empty. The scalar lets us avoid making this distinction.

As @matriv pointed out there wouldn't be a performance gain and I created a PR to improve the error message: https://github.com/crate/crate/pull/14082
